lib/bap_service/bap_service.ml

```diff
@@ -0,0 +1,103 @@
+open Core_kernel.Std
+open Bap_future.Std
+open Regular.Std
+
+type 'a named = {
+  name : string;
+  desc : string;
+  data : 'a;
+} [@@deriving bin_io, compare, sexp]
+
+type service = string named
+[@@deriving bin_io, compare, sexp]
+
+type provider = service named
+[@@deriving bin_io, compare, sexp]
+
+type product = {
+  digest    : String.Set.t;
+  providers : provider list;
+} [@@deriving bin_io, compare, sexp]
+
+module Service = struct
+  type t = service
+  [@@deriving bin_io, compare, sexp]
+
+  include Regular.Make(struct
+      type nonrec t = t [@@deriving bin_io, compare, sexp]
+      let module_name = Some "Bap.Service.Service"
+      let hash = Hashtbl.hash
+      let version = "0.1"
+      let pp fmt t = Format.fprintf fmt "%s %s" t.name t.desc
+    end)
+
+  let services = Table.create ()
+
+  let declare ~desc ~uuid name =
+    let t = {desc=desc; data=uuid; name;} in
+    Hashtbl.set services t (Stream.create ());
+    t
+
+  let provide t product =
+    let p = Hashtbl.find_exn services t in
```

Comment:
this will fail, since you're not checking that the declared services are unique.
```
let s1 = Service.declare ~desc:"" ~uuid:"" "" 
let s2 = Service.declare ~desc:"" ~uuid:"" "" 
let p = <create-product> 
provide s1 p;; 
```